src/compile/projection/parse.ts

```diff
@@ -0,0 +1,140 @@
+    (channel) => {
+      const def = encoding[channel];
+      return isFieldDef(def) && contains([LATITUDE, LONGITUDE], def.type);
+  });
+
+  if (isGeoShapeMark || isGeoPointMark) {
+    let data: VgSignalRef | string;
+    if (isGeoPointMark || (isGeoShapeMark && encoding[SHAPE])) {
+      // if there is a shape encoding or latitude / longitude encoding,
+      // the compiler added a geojson transform w/ a signal as specified
+      data = {
+        signal: model.getName('geojson')
+      };
+    } else {
+      // main source is geojson, so we can just use that
+      data = model.requestDataName(MAIN);
+    }
+
+    return new ProjectionComponent(model.projectionName(true), {
+      ...(config.projection || {}),
+      ...(specifiedProjection || {}),
+    } as Projection, [model.getSizeSignalRef('width'), model.getSizeSignalRef('height')], [data]);
+  }
+
+  return undefined;
+}
+
+function dryMerge(first: ProjectionComponent, second: ProjectionComponent): ProjectionComponent {
+  const properties = every(PROJECTION_PROPERTIES, (prop) => {
+    // neither has the poperty
+    if (!first.explicit.hasOwnProperty(prop) &&
+      !second.explicit.hasOwnProperty(prop)) {
+      return true;
+    }
+    // both have property and an equal value for property
+    if (first.explicit.hasOwnProperty(prop) &&
+      second.explicit.hasOwnProperty(prop) &&
+      JSON.stringify(first.get(prop)) === JSON.stringify(second.get(prop))) {
+      return true;
+    }
+    return false;
+  });
+
+  const size = JSON.stringify(first.size) === JSON.stringify
+    (second.size);
+  if (size) {
+    if (properties) {
+      return first;
+    } else if (JSON.stringify(first.explicit) === JSON.stringify
+      ({})) {
+      return second;
+    } else if (JSON.stringify(second.explicit) === JSON.stringify
+      ({})) {
+      return first;
+    }
+  }
+  return null;
+}
+
+function parseNonUnitProjections(model: Model): ProjectionComponent {
+  if (model.children.length === 0) {
+    return undefined;
+  }
+
+  let nonUnitProjection: ProjectionComponent;
+  const mergable = every(model.children, (child) => {
+    parseProjection(child);
+    const projection = child.component.projection;
+    if (!projection) {
+      // child layer does not use a projection
+      return true;
+    } else if (!nonUnitProjection) {
+      // cached 'projection' is null, cache this one
+      nonUnitProjection = projection;
+      return true;
+    } else {
+      const merge = dryMerge(nonUnitProjection, projection);
+      if (merge) {
+        nonUnitProjection = merge;
+      }
+      return !!merge;
+    }
+  });
+
+  // it cached one and all other children share the same projection,
+  if (nonUnitProjection && mergable) {
+    // so we can elevate it to the layer level
+    const name = model.projectionName(true);
+    const modelProjection = new ProjectionComponent(
+      name,
+      duplicate(nonUnitProjection.specifiedProjection),
```

Comment:
Do we really need to make duplicate for all of these (except `data`)?  (Since we don't really mutate them later?)